src/wallet/wallet.cpp

```diff
@@ -81,13 +81,46 @@ std::shared_ptr<CWallet> GetWallet(const std::string& name)
     return nullptr;
 }
 
+static std::mutex g_wallet_release_mutex;
+static std::condition_variable g_wallet_release_cv;
+static std::set<CWallet*> g_unloading_wallet_set;
+
 // Custom deleter for shared_ptr<CWallet>.
 static void ReleaseWallet(CWallet* wallet)
 {
-    wallet->WalletLogPrintf("Releasing wallet\n");
-    wallet->BlockUntilSyncedToCurrentChain();
-    wallet->Flush();
-    delete wallet;
+    {
+        std::unique_lock<std::mutex> lock(g_wallet_release_mutex);
+        // Assert that there was an explict UnloadWallet call.
+        assert(g_unloading_wallet_set.erase(wallet) == 1);
+    }
+    g_wallet_release_cv.notify_all();
+}
+
+void UnloadWallet(std::shared_ptr<CWallet> wallet)
+{
+    CWallet* pwallet = wallet.get();
+    {
+        std::unique_lock<std::mutex> lock(g_wallet_release_mutex);
+        assert(g_unloading_wallet_set.insert(pwallet).second);
+    }
+    UnregisterValidationInterface(pwallet);
+    // The wallet can be in use so it's not possible to explicitly unload here.
+    // Notify the unload intent so that all remaining shared pointers are
+    // released.
+    pwallet->NotifyUnload();
+    // Time to ditch our shared_ptr and then wait for ReleaseWallet call.
+    wallet.reset();
+    {
+        std::unique_lock<std::mutex> lock(g_wallet_release_mutex);
+        while (g_unloading_wallet_set.count(pwallet)) {
+            g_wallet_release_cv.wait(lock);
+        }
+    }
+    // Now it's safe to close and delete the wallet.
+    pwallet->WalletLogPrintf("Releasing wallet\n");
+    pwallet->BlockUntilSyncedToCurrentChain();
```

Comment:
Not sure about this, I think it should be called before `UnregisterValidationInterface` otherwise `SyncWithValidationInterfaceQueue` does nothing than waiting..

Reply:
This does seem unnecessary. But I'd also expect this code to calling UnregisterValidationInterface before flushing. I.e. stop requesting new notifications, then save state, instead of saving state first and then hoping new notifications don't arrive after the flush...